src/ndb/ndb_internal.c

```diff
@@ -0,0 +1,426 @@
+  chCondSignal(&wq_new_data);
+  chMtxUnlock(&wq_access);
+  return NDB_ERR_NONE;
+}
+
+enum ndb_op_code ndb_wq_get(ndb_element_metadata_t** md)
+{
+  chMtxLock(&wq_access);
+  if (NULL == wq_first) {
+    /* NOTE: chCondwaitTimeout does not care of
+     locking mutex again when timeout occurs*/
+    chCondWaitTimeout(&wq_new_data, NV_WRITE_REQ_TIMEOUT);
+    if (NULL == wq_first) {
+      *md = NULL;
+      return NDB_ERR_TIMEOUT;
+    }
+  }
+
+  *md = wq_first;
+  wq_first = wq_first->next;
+
+  if (NULL == wq_first)
+    wq_last = NULL;
+
+  chMtxUnlock(&wq_access);
+  return NDB_ERR_NONE;
+}
+
+void ndb_init()
+{
+  if (!do_writes_to_flash)
+    log_info("NDB: configured not to save data to flash file system");
+
+  u16 vlen = sizeof(GIT_VERSION);
+  vlen = vlen > sizeof(ndb_file_version) ? sizeof(ndb_file_version) : vlen;
+  memcpy(ndb_file_version, GIT_VERSION, vlen);
+  if (vlen < sizeof(ndb_file_version)) {
+    memset(&ndb_file_version[vlen], 0xb6, sizeof(ndb_file_version) - vlen);
+  }
+}
+
+void ndb_start()
+{
+  chThdCreateStatic(ndb_thread_wa, sizeof(ndb_thread_wa),
+                    NDB_THREAD_PRIORITY, ndb_service_thread, NULL);
+}
+
+void ndb_log_file_open(const enum ndb_op_code oc, const char* file_type,
+                   const ndb_file_t* file)
+{
+  switch (oc) {
+    case NDB_ERR_NONE:
+      log_info("Opened %s file (%d)", file_type, file->fh);
+      break;
+    case NDB_ERR_FILE_IO:
+      log_error("Cannot open %s file", file_type);
+      break;
+    case NDB_ERR_INIT_DONE:
+      log_info("No %s file present in flash, create an empty one (%d)",
+               file_type, file->fh);
+      break;
+    default:
+      assert(true);
+      break;
+  }
+}
+
+void ndb_load_data(ndb_file_t *f,
+                   const char* ftype,
+                   void *data,
+                   ndb_element_metadata_t* metadata,
+                   const size_t el_size,
+                   const size_t el_number)
+{
+  enum ndb_op_code r = NDB_ERR_FILE_IO;
+  size_t ds = el_size * el_number;
+  size_t mds = sizeof(ndb_element_metadata_nv_t) * el_number;
+  ndb_element_metadata_nv_t md_nv[el_number];
+
+  if (do_writes_to_flash) {
+    if (ndb_open_file(f) == NDB_ERR_NONE) {
+      if (ndb_read(f, 0, data, ds) == NDB_ERR_NONE) {
+        if (ndb_read(f, ds, md_nv, mds) == NDB_ERR_NONE) {
+          u32 i;
+          for (i = 0; i < el_number; i++) {
+            metadata[i].nv_data = md_nv[i];
+          }
+          r = NDB_ERR_NONE;
+        }
+      }
+    }
+  }
+
+  memset(data, 0, ds);
+  memset(metadata, 0, sizeof(ndb_element_metadata_t) * el_number);
+
+  if (do_writes_to_flash && (f->fh >= 0)) {
+    memset(md_nv, 0, mds);
+    if (ndb_write_file_data(f, 0, data, ds) == NDB_ERR_NONE) {
+      if (ndb_write_file_data(f, ds, md_nv, mds) == NDB_ERR_NONE) {
+        if (ndb_write_file_data(f, ds + mds,
+                                &ndb_file_end_mark, 1) == NDB_ERR_NONE) {
+          r = NDB_ERR_INIT_DONE;
```

Comment:
`r` will be set to NDB_ERR_NONE if the top block succeeds but the bottom block does not.